### ISSUE TITLE
HLSL: SPIRV-remap --dce func stripping things needed for validation?

### ISSUE THREAD

**Issue by dankbaker** [KhronosGroup/glslang#617]
When compiling this HLSL source shader, then running through spirv-remap with the --dce func (Dead code removal), when I create the shader I get this message, and similar message for almost all SPIRV files generated from HLSL, however the shaders seem to work and the app renders correctly:

```
SC (ERROR): object: 0x0 type: 0 location: 9664 msgCode: 5: SPIR-V module not valid: The following forward referenced IDs have not be defined:
15[TestShaderConsts(struct-VSOut-vf4-vf21;] 14[Input] 13[ColorShader(struct-VSOut-vf4-vf21;] 12[Input] 17[RTTestShader(struct-VSOut-vf4-vf21;] 16[Input] 24[PosShaderInstance(u1;u1;struct-VSOut-vf4-vf21;] 21[VertexID] 22[Instance] 23[OutVertex] 26[InstancePS(struct-VSOut-vf4-vf21;] 25[Input] 90[RenderColor] 121[vPos] 122[vTex]
```

This is the source HLSL file, with vert shader entry point  PosShader:
```
#pragma warning( disable : 3078 )
 #pragma warning( disable : 3571 )
 #pragma warning( disable : 3550 )
 #pragma warning( disable : 4008 )
 #if _D3D_ == 1 
#define layout(a,b)
#else
#define register(a) blank
#endif

layout(set=9,binding=0) sampler SS_DEFAULT       : register(s0);
layout(set=9,binding=1) sampler SS_DEFAULT_CLAMP : register(s1);
layout(set=9,binding=2) sampler SS_ANISO         : register(s2);
layout(set=9,binding=3) sampler SS_ANISO_CLAMP   : register(s3);
layout(set=9,binding=4) sampler SS_POINT         : register(s4);
layout(set=9,binding=5) sampler SS_POINT_CLAMP   : register(s5);

layout(set=0,binding=0) cbuffer ShaderDynamics : register(b0)
{
    float4 s_Color : packoffset(c0.x);
    float4 s_Color2 : packoffset(c1.x);
    float2 Scale : packoffset(c2.x);
    float2 Offset : packoffset(c2.z);
    float ZDepth : packoffset(c3.x);
    float2 TexCoord : packoffset(c3.y);
    uint uBufferIndex : packoffset(c3.w);
    uint uBufferValue : packoffset(c4.x);
    float4 DynamicTest[3] : packoffset(c5.x);
};

//ResourceSet VertexData 
    layout(set=1,binding=0)Buffer<float4> V_Position : register(t0);
    layout(set=1,binding=1)Buffer<float4> V_TexCoord : register(t1);
//EndTexture Set

layout(set=0,binding=3) cbuffer ShaderConsts : register(b3)
{
    float4 s_ShaderColor : packoffset(c0.x);
};

//ResourceSet PixelData 
    layout(set=2,binding=0)Texture2DArray TestTexture : register(t16);
    layout(set=2,binding=1)Texture2D TestRTTexture : register(t17);
    layout(set=2,binding=2)Texture2D<float4> TestUAVTexture : register(t18);
    layout(set=2,binding=3)Buffer<uint> TestUAVBuffer : register(t19);
//EndTexture Set
#line  40 "D:/p4/OxideHaswell/Main/Apps/GraphicsTest/TestShaders.oxsl"
    // -----------------------------------------------------------------------
    struct VSOut
    {
        float4 fPos    : SV_POSITION;
		float2 fTex	   : TEXCOORD;
    };


    // -----------------------------------------------------------------------
    void PosShader( uint VertexID : SV_VertexID, out VSOut OutVertex )
    {
        float4 vPos = float4( V_Position.Load( (int) VertexID ).xyz + ZDepth, 1 );
		float2 vTex = float2( V_TexCoord.Load( (int) VertexID ).xy);
		OutVertex.fPos = vPos;
		OutVertex.fPos.xy = vPos.xy * Scale + Offset;
		OutVertex.fTex = vTex;       
    }


    float4 ColorShader( VSOut Input ) : SV_TARGET0
    {
		float4 RenderColor = TestRTTexture.Sample(SS_DEFAULT, Input.fTex);
		//return lerp(s_Color * TestTexture.Sample(SS_DEFAULT, float3(Input.fTex,1)), RenderColor, RenderColor.a);
		//if(any(!isfinite(s_ShaderColor)))
		//	return float4(1,0,1,1);
		return s_ShaderColor + TestTexture.Sample(SS_DEFAULT, float3(Input.fTex, 1))*.01;
    }

	
    float4 TestShaderConsts( VSOut Input ) : SV_TARGET0
    {
		return s_ShaderColor; 
	}

	float4 RTTestShader( VSOut Input ) : SV_TARGET0
    {

		if(TestUAVBuffer.Load((int)uBufferIndex.x).x == uBufferValue)
			return float4(1,0,1,1);
		else
			return float4(0,0,1,1);

    }

	// -----------------------------------------------------------------------
    void PosShaderInstance( uint VertexID : SV_VertexID, uint Instance : SV_InstanceID, out VSOut OutVertex)
    {
    

        float4 vPos = float4( V_Position.Load( (int) VertexID ).xyz + ZDepth, 1 );
		float2 vTex = float2( V_TexCoord.Load(  (int) VertexID ).xy);
		OutVertex.fPos = vPos;
		OutVertex.fPos.xy = vPos.xy * Scale + Offset * float(Instance) + float2(-1,-1);
		OutVertex.fTex = vTex;
    }

	
	float4 InstancePS( VSOut Input ) : SV_TARGET0
    {
		//This works, but the other line does not, even though uBufferIndex.x == 0
		//return DynamicTest[0];

		return TestTexture.Sample(SS_DEFAULT, float3(Input.fTex, 2));
    }

```

**Comment by johnkslang** [comment]
@steve-lunarg is this a remapper bug?

**Comment by steve-lunarg** [comment]
Seems likely so, although I don't quite understand it yet.  Can I duplicate this using spirv-val?  e.g, something like:

```
$ glslangValidator -S frag -i -H -V -D -e InstancePS ~/hlsl/forward-ref.hlsl
$ spirv-remap -v --dce func -i frag.spv -o /tmp/remapped/
$ spirv-val /tmp/remapped/frag.spv

```

**Comment by dankbaker** [comment]
Where do I get spirv-val to try that? I basically see validation errors on essentially every HLSL generated shader that we run through spirv-remap. but no errors if I don't use the util. The util strips off tons of dead code though so it's not much option for us to not use it.

**Comment by steve-lunarg** [comment]
spirv-val is part of the [SPIRV-Tools](https://github.com/KhronosGroup/SPIRV-Tools) project.  It's a validator that can report on problems with SPIR-V modules.

I've just tried it on the result as above but it didn't report anything.  I'll guess this is a problem with the remapper not removing some references to functions it eliminated.  In any case, I'll look into this more today, although it involves some guesswork without a way for me to duplicate the symptoms.

**Comment by steve-lunarg** [comment]
@dankbaker - Are you using the --strip all option?

I've looked at the SPV from the above shader and if I leave debug info in, the spirv-remap tool does not remove the OpNames that mention the now-deleted symbols.  It should, so that's a bug, which I'll make a patch to fix.  However if you still see the symptom even when stripping debug info, then what I'm looking at isn't _your_ bug, but merely _another_ bug, so I'll need to look harder.

**Comment by dankbaker** [comment]
No, this is what I'm using: "spirv-remap.exe --dce funcs -i %s -o ." --dce all I believe actually crashes spriv-remap, though I think I got yet more messages with the --strip all.

It isn't blocking me, as the shaders seem to work and I set up the error message reporting to ignore the error.

**Comment by steve-lunarg** [comment]
@dankbaker: I fixed the debug opcode problem in WIP PR #624.  See how you get on with that.

I manually searched the result of a "--dce all" run on your example shader for remaining instances of removed IDs, and didn't find any.  If you still see that message, then I'll need to look for an additional cause.

I also ran a --dce all pass on a large (~50Mb) shader collection with no problems.  If you see a crash on something and can send me the shader, I'll check that out.

Separately, we may want to file a request for spirv-val to catch cases about references to non-present IDs.
